omniscidb/QueryEngine/Execute.cpp

```diff
@@ -1360,6 +1361,76 @@ ResultSetPtr Executor::reduceSpeculativeTopN(
+    const QueryMemoryDescriptor& query_mem_desc,
+    std::shared_ptr<RowSetMemoryOwner> row_set_mem_owner) const {
+  std::vector<ResultSetPtr> results;
+  std::vector<int64_t*> buffers;
+
+  results.reserve(results_per_device.size() + 1);
+  buffers.reserve(results_per_device.size() + 1);
+
+  // In the reduction result we want each fragment result to hold an offset in
+  // output buffer instead of number of rows. Also, we want to make an additional
+  // result set holding final partition sizes. Achieve it by using a new zero-filled
+  // buffer for the the first fragment and partial sums for all other having
+  // partition sizes in the last of them.
+  // Additionally, we want them to be columnar Projection instead of PerfectHash
+  // to later zero-copy fetch all rows.
+  using IndexedResultSet = std::pair<ResultSetPtr, std::vector<size_t>>;
+  std::sort(results_per_device.begin(),
+            results_per_device.end(),
+            [](const IndexedResultSet& lhs, const IndexedResultSet& rhs) {
+              CHECK_GE(lhs.second.size(), size_t(1));
+              CHECK_GE(rhs.second.size(), size_t(1));
+              return lhs.second.front() < rhs.second.front();
+            });
+
+  auto parts = query_mem_desc.getEntryCount();
+  auto proj_mem_desc = query_mem_desc;
+  proj_mem_desc.setQueryDescriptionType(QueryDescriptionType::Projection);
+  proj_mem_desc.setHasKeylessHash(false);
+  proj_mem_desc.clearGroupColWidths();
+  // Currently, all perfect hash tables are expected to to use 8 byte padded width.
+  CHECK_EQ(static_cast<int>(proj_mem_desc.getPaddedSlotWidthBytes(0)), 8);
+  auto first_rs =
+      std::make_shared<ResultSet>(results_per_device.front().first->getTargetInfos(),
+                                  ExecutorDeviceType::CPU,
+                                  proj_mem_desc,
+                                  row_set_mem_owner,
+                                  data_mgr_,
+                                  blockSize(),
+                                  gridSize());
+  first_rs->allocateStorage(plan_state_->init_agg_vals_);
+  results.push_back(first_rs);
+  buffers.push_back(
+      reinterpret_cast<int64_t*>(first_rs->getStorage()->getUnderlyingBuffer()));
+  for (auto& pr : results_per_device) {
+    auto buf = pr.first->getStorage()->getUnderlyingBuffer();
+    auto proj_rs =
+        std::make_shared<ResultSet>(results_per_device.front().first->getTargetInfos(),
+                                    ExecutorDeviceType::CPU,
+                                    proj_mem_desc,
+                                    row_set_mem_owner,
+                                    data_mgr_,
+                                    blockSize(),
+                                    gridSize());
+    proj_rs->allocateStorage(buf, {});
+    results.push_back(proj_rs);
+    buffers.push_back(reinterpret_cast<int64_t*>(buf));
+  }
+
+  bzero(buffers[0], sizeof(int64_t) * parts);
+  for (size_t i = 2; i < buffers.size(); ++i) {
```

Comment:
Why does this index start at 2?

Reply:
The first one is all zeroes, so no need to add them to the second one, and therefore start from the third one.